### ISSUE TITLE
Idea: `our-picture` for complicated PICTURE and SOURCE elements

### ISSUE THREAD

**Issue by markadrake** [umbraco-community/Our-Umbraco-TagHelpers#29]
I often write a dummy version of this across all my projects. I'm unfamiliar with Tag Helpers but I'm going to spend a little time tomorrow looking into how difficult this would be. It seems like it would be helpful to me at least.

Is there any interest in creating a tag helper for output of complicated `<picture>` elements? The ideal result is a `<picture>` element that can target all of these scenarios:

- Multiple media queries, and
- Multiple image crop aliases, and
- Multiple file types

## **Proposed Input**

I'm new to the tag helpers concept in .NET Core. It seems that this is complex enough that it may require a model first?

```
OurPicture myPictureModel = new OurPicture {

     // by media query
     "screen and (max-width: 1023px)": {
          media: myIPublishedContent,
          formats: ["png", "webp"],
          cropAlias: "myCropAlias"
     },

     "screen and (min-width: 1024px)": {
          media: myIPublishedContent,
          formats: ["png", "webp"],
          cropAlias: "myCropAlias2"
     }
};
...
<picture our-picture="myPictureModel">
     <img src="/path/to/myIPublishedContent" alt="myAltValue" />
</picture>
```

## **Proposed Output**

As you can see below, this becomes painstaking to write manually. I often begin my responsive design with 5 target breakpoints (xs, s, m, l, xl). This example below is only using 2.

```
<picture>
     <source media="screen and (max-width: 1023px)"
          type="image/png"
          srcset="/path/to/myIPublishedContent?format=png">
     <source media="screen and (max-width: 1023px)"
          type="image/webp"
          srcset="/path/to/myIPublishedContent?format=png">
     <source media="screen and (min-width: 1024px)"
          type="image/png"
          srcset="/path/to/myIPublishedContent?format=webp">
     <source media="screen and (min-width: 1024px)"
          type="image/webp"
          srcset="/path/to/myIPublishedContent?format=webp">
     <img src="/path/to/myIPublishedContent" alt="myAltValue" />
</picture>
```

## **References**

MDN reference for `PICTURE` element: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/picture

**Comment by markadrake** [comment]
Got a start on building something like this.

![image](https://user-images.githubusercontent.com/713355/150478224-12e073d6-78b9-4cca-aa8a-f3eec8635733.png)

Got the HTML output looking good.

![image](https://user-images.githubusercontent.com/713355/150478260-dced2258-7052-40ac-a324-a68ef84dc824.png)

I just need to take advantage of the `image crop alias`, and appending `?format=myFormat` to the resulting url and I got something useful here!

I'm thinking about adding a set of default values so `Media`, `ImageFormats`, and `ImageCropAlias` don't have to be redefined for each media query.

**Comment by markadrake** [comment]
Just another update, mostly for me so I know where I left off with this ;)

I created default options, which serve to create the IMG element as well as serve as the default values for each SOURCE element.

![image](https://user-images.githubusercontent.com/713355/150490208-72c524c5-c551-4666-b7e7-116b8b0741c6.png)

**Comment by warrenbuckley** [comment]
Sorry @markadrake I have been pretty bad on keeping top of all my little pet project repos 🙈 
@AndyBoot has done a PR here https://github.com/umbraco-community/Our-Umbraco-TagHelpers/pull/44

Perhaps take a look at Andy's work to see if it suits your needs.

**Comment by warrenbuckley** [comment]
Closed with @AndyBoot PR in #44
